lib/change_stream.js

```diff
@@ -506,46 +495,26 @@ function processNewChange(args) {
       // close internal cursor, ignore errors
       changeStream.cursor.close();
 
-      // attempt recreating the cursor
-      if (eventEmitter) {
-        waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
-          if (err) {
+      waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
+        if (err) {
+          changeStream.closed = true;
+          if (eventEmitter) {
             changeStream.emit('error', err);
             changeStream.emit('close');
             return;
           }
-          changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
-        });
-
-        return;
-      }
-
-      if (callback) {
-        waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
-          if (err) return callback(err, null);
-
-          changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
-          changeStream.next(callback);
-        });
+          return callback(err, null);
+        }
 
-        return;
-      }
-
-      return new Promise((resolve, reject) => {
-        waitForTopologyConnected(topology, { readPreference: options.readPreference }, err => {
-          if (err) return reject(err);
-          resolve();
-        });
-      })
-        .then(
-          () => (changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions))
-        )
-        .then(() => changeStream.next());
+        changeStream.cursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
+        if (eventEmitter) return;
+        changeStream.next(callback);
```

Comment:
I think we still have a bug here if the thing that causes the recreate is `hasNext` (I think we discussed this the other day), I think a safer replacement would be to call `hasNext` here since that should cache the results.

Reply:
I was planning on deferring any logical changes to `ChangeStream` unrelated to the new `closed` property to an immediate follow-up PR to this, for [NODE-2548](https://jira.mongodb.org/browse/NODE-2548). Since `processNewChange` was using `next` before, I'd like to keep that for this PR; changing that to `hasNext` isn't in this one's scope. LMK if you disagree!